### PR TITLE
Support Format NVM and Queue Deletion

### PR DESCRIPTION
**Pull request by csm10495**

### CODE REVIEW

File: cNVMe/Controller.cpp [hunk 1]
```diff
@@ -551,6 +554,119 @@ namespace cnvme
 			LOG_INFO("Held onto submission queue with an id of " + std::to_string(command.DW10_CreateIoQueue.QID));
 		}
 
+		NVME_CALLER_IMPLEMENTATION(adminDeleteIoCompletionQueue)
+		{
+			Queue* q = this->getQueueWithId(this->ValidCompletionQueues, command.DW10_DeleteIoQueue.QID);
+
+			// You can't delete my admin queue!
```
Comment (mfamis) [inline cNVMe/Controller.cpp:595]: 10/10 documentation

File: cNVMe/Controller.cpp [hunk 2]
```diff
@@ -551,6 +554,119 @@ namespace cnvme
...
+			std::set<UINT_32> namespacesToFormat;
+			bool shouldFormatAll = (command.NSID == ALL_NAMESPACES);
+
+			// Should format all if IC says format all on crypto so and this is a crypto erase
+			shouldFormatAll |= (command.DW10_Format.SES == constants::commands::format::ses::CRYPTOGRAPHIC_ERASE && this->IdentifyController.AllNamespacesErasedOnSecureErase);
+
+			// Should format all if IC says format all on namespaces so and this is not a crypto erase
+			shouldFormatAll |= (command.DW10_Format.SES != constants::commands::format::ses::CRYPTOGRAPHIC_ERASE && this->IdentifyController.FormatAppliesToAllNamespaces);
+
+			// If we should format all namespaces, make sure we have all of them
```
Comment (mfamis) [inline cNVMe/Controller.cpp]: This comment needs to be clarified, as "make sure we have all of them" sounds like we are verifying the presence of the namespaces. Something along the lines of: "If we should format all of the namespaces, add all of the registered namespaces to the set of namespaces to be formatted."

File: cNVMe/Controller.cpp [hunk 3]
```diff
@@ -551,6 +554,119 @@ namespace cnvme
...
+			}
+
+			// Call format on all namespacesToFormat... if any fail... fail the command
+			for (auto &nsid : namespacesToFormat)
+			{
+				completionQueueEntryToPost = this->NamespaceIdToNamespace[nsid].formatNVM(command);
+
+				if (completionQueueEntryToPost.SC != 0)
+				{
+					break; // make sure we leave this loop now.
```
Comment (mfamis) [inline cNVMe/Controller.cpp]: Is this a fail condition? Do we need to log something here?

File: cNVMe/Tests.cpp [hunk 4]
```diff
@@ -268,9 +269,82 @@ namespace cnvme
...
+				UINT_32 timeout = 5; // arbitrary
+				pDriverCommand->Timeout = timeout;
+				pDriverCommand->TransferDataDirection = cnvme::driver::NO_DATA;
+				pDriverCommand->Command.DWord0Breakdown.OPC = constants::opcodes::admin::DELETE_IO_COMPLETION_QUEUE;
+				pDriverCommand->Command.DW10_DeleteIoQueue.QID = ADMIN_QUEUE_ID;
+
+				driver.sendCommand(buffer, BUF_SIZE);
+
+				auto status = pDriverCommand->CompletionQueueEntry.SC;
+				ASSERT_IF(status != constants::status::codes::specific::INVALID_QUEUE_IDENTIFIER, "Controller did not fail deleting the admin queue correctly");
```
Comment (mfamis) [inline cNVMe/Tests.cpp]: Maybe reword this to clarify the error: "Expected controller to fail IO queue deletion with admin queue ID, but did not receive INVALID_QUEUE_IDENTIFIER status."

File: cNVMe/Tests.h [hunk 5]
```diff
@@ -102,6 +102,11 @@ namespace cnvme
 			/// Tests that the controller can fail a command with an unknown command opcode
 			/// </summary>
 			bool testNVMeCommandOpcodeInvalid();
+
+			/// <summary>
+			/// Tests that if we create an SQ/CQ we have to delete the SQ first
+			/// </summary>
+			bool testNVMeQueueDeletionFailures();
```
Comment (mfamis) [inline cNVMe/Tests.h:109]: Is there already a formatNVM test?

**Review by mfamis — APPROVED** [review]
I added some recommendations on logging/documentation changes, but the code changes look great.

File: cNVMe/Controller.cpp (hunk 3 above)
Comment (csm10495) [reply cNVMe/Controller.cpp]: It is. Though the upper level should set the completionQueueEntryToPost so we'll just return up. Should be ok.

File: cNVMe/Tests.h (hunk 5 above)
Comment (csm10495) [reply cNVMe/Tests.h:109]: It's tough to test now because there is no read/write. Holding off for now.

**Comment by csm10495** [comment]
Merged. Closes #17 and closes #18 .
